### ISSUE TITLE
[member-ordering] public member reference private member should be ignored

### ISSUE THREAD

**Issue by JounQin** [typescript-eslint/typescript-eslint#1166]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/member-ordering": 2
  }
}
```

```TS
export class A {
  private a: number
  private b: number

  c = this.a + this.b
}
```

**Expected Result**

Do not report error

**Actual Result**

Unfixable

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.6.0` |
| `@typescript-eslint/parser`        | `2.6.0` |
| `TypeScript`                       | `3.6.4` |
| `ESLint`                           | `6.6.0` |
| `node`                             | `12.13.0` |
| `npm`                              | `6.12.0` |

**Comment by bradzacher** [comment]
same as #417

Detecting the dependencies between fields is trivial, but using that to inform how the sort works is very complex.

For example:

```ts
export class A {
  private a: number
  private b: number
  private d: number

  protected e: number

  c = this.a + this.b
}
```

where does `c` go?
- does it go directly below `a` and `b`, which is the highest available place which also satisfies the dependency constraints?
- does it go after `d`, which is the highest available place after a block (i.e. doesn't interrupt the private modifier group) that also satisfies the dependency constraint?
- does it go at the end of all "field" type members (i.e. after `e`), so that it's a lonely public field?
- does it get ignored completely, and we leave it up to the dev?

What about if you introduce another constraint:

```ts
export class A {
  private a: number
  private b: number
  private d: number

  protected e: number

  c = this.a + this.b
  private f = this.c
}
```

now what happens to `f`? where does it go?

It's not always easy to detect when there is a dependency as well - there are a lot of cases that are irrelevant or just hard to detect via the AST.

```ts
abstract class Foo {
    a = 1;
   
    abstract boo: number;
    get aaa() { return this.boo }
}
class Bar extends Foo {
    // irrelevant dependency on parent field
    b = this.a;
    
    // irrelevant dependency on a class method
    c = this.myMethod;
    myMethod() {}
    
    get foo() { return this.b }
    // hard to detect dependency on this.b
    bar = this.foo

    boo = 1;
    // impossible to detect dependency on this.boo
    baz = this.aaa
}
```

Dependency based ordering constraints add a lot of complexity into an already complex rule (which is also getting more complex in #263).

---

The suggestion is to either use a disable comment, or use a constructor to completely avoid this problem:
```ts
export class A {
  c: number;

  private a: number
  private b: number
  private d: number
  private f: number

  protected e: number

  constructor() {
    this.c = this.a + this.b
    this.f = this.c
  }
}
```

**Comment by JounQin** [comment]
@bradzacher Why can't we just add an option to ignore this case or by default?

It is something just related to code style, I don't think it's worth to change my source code into constructor only due to this.

`a disable comment` is a bad idea if we have a lot of public member relied on private member(s).
